### PR TITLE
Initial tests for Dynamic Imports

### PR DESCRIPTION
**Pull request by rwaldron**
~~This is not ready for in depth review yet~~

Ref #1164

### CODE REVIEW

**Comment by dnalborczyk** [comment]
hey guys, sorry for hijacking this PR.

we're trying to make `esm` https://github.com/standard-things/esm (a node module supporting `ES modules`, `dynamic import`, and also a `bridge to commonjs`) ECMA-262 compliant as much as possible, and we are running a selection of tests, mainly everything from `module-code`, `import` and `export`.

we also started running against this branch/PR to get some test coverage for `dynamic import`, although we know it might be not in it's final stages, and hasn't been reviewed yet.

there is a couple tests involving `import('')`, and I was wondering how the result should be evaluated? 
meaning, what determines the test to be successful? since `browsers`, and `node --experimental-modules`, as well as `esm` return a rejected Promise not being able to find a module.

for example: [module-code/dynamic-import/syntax-nested-arrow-assignment-expression-empty-str-arg.js](https://github.com/tc39/test262/pull/1588/commits/269908b039a11e806b5e9fe796424b7b7687f2d2#diff-35e14829fe8c492ca366117539c56ffb)

any hints or pointers are much appreciated. thank you in advance!

**Comment by ljharb** [comment]
Since it’s a URL, there’s no reason that an empty string necessarily won’t resolve to a module.

**Comment by leobalter** [comment]
Rick is out until September and I didn't have the chance to review this PR yet. I'm not sure at what point this PR is still a WIP with parts not done yet and I'm afraid I could be over reviewing things Rick already has plans to change.

My current work efforts are based on class fields related stuff but I hope we can be back to this PR anytime.

**Comment by ljharb** [comment]
Module loading is engine-specific and it’d be very hard to cover this fully with test262.

**Comment by rwaldron** [comment]
> Module loading is engine-specific and it’d be very hard to cover this fully with test262.

Presently we use engine-specific machinery, provided via eshost runtime definitions, to ensure that module loading operations occur within the scope of each host's capabilities. This is how both module code and dynamic import test material is executed.

**Comment by leobalter** [comment]
I just rebased the branch to verify the CI again. No other changes introduced here so far.

**Comment by rwaldron** [comment]
@dnalborczyk I think your question is more appropriate for the proposal/authors. 

> Meanwhile I might just blacklist those tests for the time being, or find another temporary workaround, e.g. parsing for import(''), and expect a rejected Promise.

"parsing for import('')", and evaluating what it returns are two very distinct aspects of the feature, which occur during different phases of a JS program's lifetime.

File: src/dynamic-import/empty-args-early-error.case [hunk 1]
```diff
@@ -3,17 +3,12 @@
 // This code is governed by the BSD license found in the LICENSE file.
 /*---
 desc: It's a SyntaxError if AssignmentExpression is omitted
-template: syntax
-negative:
-  phase: parse
-  type: SyntaxError
+template: syntax/invalid
```
Comment (rwaldron) [inline src/dynamic-import/empty-args-early-error.case]: Nice!

**Comment by leobalter** [comment]
I finally fixed the duplicate tests, they should pass on the CI process now.

`git grep -h -E 'path\: (.*)' src/dynamic-import/ | uniq -c` should be helpful as a future reference

**Comment by jdalton** [comment]
@leobalter The file https://github.com/tc39/test262/blob/dynamic-import/test/language/module-code/dynamic-import/catch/nested-arrow-import-catch-instn-iee-err-ambiguous-import.js is trying to import a missing file (maybe a path mixup)

**Comment by leobalter** [comment]
Thanks for catching this up. This is still WIP but you definitely caught some commit mixup. I’ll restore this file tomorrow.

Please, Let me know if you find anything else from the tests. Looking forward to complete this coverage.

**Comment by jdalton** [comment]
Np. There's a couple more to investigate for missing files:

```

# DYNAMIC IMPORT ❌ (incomplete test)
test/language/module-code/dynamic-import/catch/nested-async-function-await-instn-iee-err-ambiguous-import.js

# DYNAMIC IMPORT ❌ (incomplete test)
test/language/module-code/dynamic-import/catch/nested-async-function-instn-iee-err-ambiguous-import.js

# DYNAMIC IMPORT ❌ (incomplete test)
test/language/module-code/dynamic-import/catch/nested-async-function-return-await-instn-iee-err-ambiguous-import.js

# DYNAMIC IMPORT ❌ (incomplete test)
test/language/module-code/dynamic-import/catch/nested-block-import-catch-instn-iee-err-ambiguous-import.js

# DYNAMIC IMPORT ❌ (incomplete test)
test/language/module-code/dynamic-import/catch/nested-block-labeled-instn-iee-err-ambiguous-import.js

# DYNAMIC IMPORT ❌ (incomplete test)
test/language/module-code/dynamic-import/catch/nested-do-while-instn-iee-err-ambiguous-import.js

# DYNAMIC IMPORT ❌ (incomplete test)
test/language/module-code/dynamic-import/catch/nested-else-import-catch-instn-iee-err-ambiguous-import.js

# DYNAMIC IMPORT ❌ (incomplete test)
test/language/module-code/dynamic-import/catch/nested-function-import-catch-instn-iee-err-ambiguous-import.js

# DYNAMIC IMPORT ❌ (incomplete test)
test/language/module-code/dynamic-import/catch/nested-if-import-catch-instn-iee-err-ambiguous-import.js

# DYNAMIC IMPORT ❌ (incomplete test)
test/language/module-code/dynamic-import/catch/nested-while-import-catch-instn-iee-err-ambiguous-import.js

# DYNAMIC IMPORT ❌ (incomplete test)
test/language/module-code/dynamic-import/catch/top-level-import-catch-instn-iee-err-ambiguous-import.js
```

**Comment by leobalter** [comment]
@jdalton this should be fixed for now

**Comment by leobalter** [comment]
ok, with the latest commits I'm able to run all the tests just OK w/ JSC and V8.

```
[~/dev/test262] dynamic-import
test262-harness --hostType=jsc --hostPath=`which jsc` -- 'test/language/module-code/dynamic-import/**/*'
Ran 272 tests
272 passed
0 failed
[~/dev/test262] dynamic-import
test262-harness --hostType=d8 --hostPath=`which v8` -- 'test/language/module-code/dynamic-import/**/*'
Ran 272 tests
272 passed
0 failed
```

**Comment by leobalter** [comment]
I'm also migrating the checklist to #1164

**Comment by leobalter** [comment]
Everything should be all set for an initial merge and get feedback from implementations while we continue working on the missing coverage found (see #1164).

@rwaldron ok to land this from your side?

**Comment by rwaldron** [comment]
Just a note to `test262-harness` users: you MUST be upgraded to `>=test262-harness@4.5.0` for these tests.

File: test/language/module-code/dynamic-import/eval-self-once-module.js [hunk 2]
```diff
@@ -11,8 +11,11 @@ info: |
        a. Let requiredModule be ? HostResolveImportedModule(module, required).
        b. Perform ? requiredModule.ModuleEvaluation().
     [...]
+
+    This test is meant to be flagged as module code, it should not initially
+    run as script code or the result will not be the same.
 includes: [fnGlobalObject.js]
-flags: [async]
+flags: [async, module]
```
[the remark this answers is not in the document]
Comment (leobalter) [reply test/language/module-code/dynamic-import/eval-self-once-module.js]: Sorry for that, I left the changes in the editor without saving to the file and pushed the other commit.

File: test/language/module-code/dynamic-import/eval-self-once-script.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+Promise.all([
+  import('./eval-self-once-script.js'),
+  import('./eval-self-once-script.js'),
+]).then(async () => {
+  // Use await to serialize imports
+  await import('./eval-self-once-script.js');
+  await import('./eval-self-once-script.js');
+
+  assert.sameValue(global.evaluated, 2, 'global property was defined and incremented only once');
+}).then($DONE, $DONE);
```
[the remark this answers is not in the document]
Comment (leobalter) [reply test/language/module-code/dynamic-import/eval-self-once-script.js]: This is intended to be this way. It's a very similar test but it is evaluated as a script and it loads itself again following this:

```
Success path

  The completion value of any subsequent call to HostResolveImportedModule after
  FinishDynamicImport has completed, given the arguments referencingScriptOrModule
  and specifier, must be a module which has already been evaluated, i.e. whose
  Evaluate concrete method has already been called and returned a normal completion.
```

you can find this normative text here: https://tc39.github.io/proposal-dynamic-import/#sec-hostimportmoduledynamically

a script file loading it self ends up being evaluated twice. The first time it was not evaluated as a Module Record or using HostResolveImportedModule. This is tricky.

[the remark this answers is not in the document]
Comment (leobalter) [reply test/language/module-code/dynamic-import/eval-self-once-script.js]: Also, the `info` in the metadata says the file is meant to run as a script code (strict or non strict) and not as a module code, which has a different evaluation.

[the remark this answers is not in the document]
Comment (jdalton) [reply test/language/module-code/dynamic-import/eval-self-once-script.js]: The description of the assert should probably be changed
> 'global property was defined and incremented only once'

[the remark this answers is not in the document]
Comment (leobalter) [reply test/language/module-code/dynamic-import/eval-self-once-script.js]: yes, thanks!

File: test/language/module-code/dynamic-import/usage/nested-if-braceless-eval-gtbndng-indirect-update-dflt.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+       a. Let M and N2 be the indirection values provided when this binding for
+          N was created.
+       b. Let targetEnv be M.[[Environment]].
+       c. If targetEnv is undefined, throw a ReferenceError exception.
+       d. Let targetER be targetEnv's EnvironmentRecord.
+       e. Return ? targetER.GetBindingValue(N2, S).
+
+---*/
+
+if (true) import('./eval-gtbndng-indirect-update-dflt_FIXTURE.js');
```
Comment (jdalton) [inline test/language/module-code/dynamic-import/usage/nested-if-braceless-eval-gtbndng-indirect-update-dflt.js]: ☝️ Based on [INTERPRETING](https://github.com/tc39/test262/blob/377f07f502890284d1e89ecb6b66343464fe14c0/INTERPRETING.md) async flagged tests should print `'Test262:AsyncTestComplete'` or `'Test262:AsyncTestFailure: ' + reason` by way of [`doneprintHandle`](https://github.com/tc39/test262/blob/master/harness/doneprintHandle.js). Without print a complete or fail the interpretation can hit a test timeout and interpret it as a failure.

File: src/dynamic-import/default/nested-if-braceless.template [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  7. IfAbruptRejectPromise(specifierString, promiseCapability).
+  8. Perform ! HostImportModuleDynamically(referencingScriptOrModule, specifierString, promiseCapability).
+  9. Return promiseCapability.[[Promise]].
+
+
+features: [dynamic-import]
+flags: [async]
+---*/
+
+if (true) import(/*{ params }*/);
```
Comment (jdalton) [inline src/dynamic-import/default/nested-if-braceless.template]: *Note: Posting on a file not hidden by GitHub.*

☝️ Based on [INTERPRETING](https://github.com/tc39/test262/blob/377f07f502890284d1e89ecb6b66343464fe14c0/INTERPRETING.md) async flagged tests should print `'Test262:AsyncTestComplete'` or `'Test262:AsyncTestFailure: ' + reason` by way of [`doneprintHandle`](https://github.com/tc39/test262/blob/master/harness/doneprintHandle.js). Without print a complete or fail the interpretation can hit a test timeout and interpret it as a failure.

File: test/language/module-code/dynamic-import/eval-self-once-script.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (rwaldron) [reply test/language/module-code/dynamic-import/eval-self-once-script.js]: @jdalton thanks for providing such quick feedback!

File: test/language/module-code/dynamic-import/usage/nested-if-braceless-eval-gtbndng-indirect-update-dflt.js (hunk 4 above)
Comment (rwaldron) [reply test/language/module-code/dynamic-import/usage/nested-if-braceless-eval-gtbndng-indirect-update-dflt.js]: Nice catch! I think these are all from my original batch of generated tests, which definitely need work. Please keep the feedback and review coming <3

**Pull request merged**